service/task_manager_module.cc

```diff
@@ -0,0 +1,160 @@
+                co_return tasks::virtual_task_hint{
+                    .table_id = table,
+                };
+            }
+        }
+    }
+    co_return std::nullopt;
+}
+
+future<tasks::is_abortable> tablet_virtual_task::is_abortable(tasks::virtual_task_hint hint) const {
+    return make_ready_future<tasks::is_abortable>(tasks::is_abortable::yes);
+}
+
+future<std::optional<tasks::task_status>> tablet_virtual_task::get_status(tasks::task_id id, tasks::virtual_task_hint hint) {
+    utils::chunked_vector<locator::tablet_id> tablets;
+    co_return co_await get_status_helper(id, tablets, std::move(hint));
+}
+
+future<std::optional<tasks::task_status>> tablet_virtual_task::wait(tasks::task_id id, tasks::virtual_task_hint hint) {
+    auto table = get_table_id(hint);
+
+    utils::chunked_vector<locator::tablet_id> tablets;
+    auto status = co_await get_status_helper(id, tablets, std::move(hint));
+    if (!status) {
+        co_return std::nullopt;
+    }
+
+    co_await _ss._topology_state_machine.event.wait([&] {
+        auto& tmap = _ss.get_token_metadata().tablets().get_tablet_map(table);
+        return std::all_of(tablets.begin(), tablets.end(), [&] (const locator::tablet_id& tablet) {
+            return tmap.get_tablet_info(tablet).repair_task_info.tablet_task_id.uuid() != id.uuid();
+        });
+    });
+
+    status->state = tasks::task_manager::task_state::done; // Failed repair task is retried.
+    status->end_time = db_clock::now(); // FIXME: Get precise end time.
+    co_return status;
+}
+
+future<> tablet_virtual_task::abort(tasks::task_id id, tasks::virtual_task_hint hint) noexcept {
+    auto table = get_table_id(hint);
+    co_await _ss.del_repair_tablet_request(table, locator::tablet_task_id{id.uuid()});
+}
+
+future<std::vector<tasks::task_stats>> tablet_virtual_task::get_stats() {
+    std::vector<tasks::task_stats> res;
+    auto tables = get_table_ids();
+    for (auto table : tables) {
+        auto& tmap = _ss.get_token_metadata().tablets().get_tablet_map(table);
+        auto schema = _ss._db.local().get_tables_metadata().get_table(table).schema();
+        std::unordered_map<tasks::task_id, tasks::task_stats> user_requests;
+        std::unordered_map<tasks::task_id, size_t> sched_num_sum;
+        co_await tmap.for_each_tablet([&] (locator::tablet_id tid, const locator::tablet_info& info) {
+            if (!info.repair_task_info.is_valid()) {
+                return make_ready_future();
+            }
+
+            auto task_id = tasks::task_id{info.repair_task_info.tablet_task_id.uuid()};
+            tasks::task_stats stats{
+                .task_id = task_id,
+                .type = locator::tablet_task_type_to_string(info.repair_task_info.request_type),
+                .kind = tasks::task_kind::cluster,
+                .scope = info.repair_task_info.is_user_request() ? "table" : "tablet",
```

Comment:
User request always means entire table is repaired? I thought we allow for repairing a single tablet too.

Reply:
We can repair a single tablet. User request specifies a list of tokens to repair. It is not always the whole table, but it is at most one table, and hence the scope

Reply:
Ok.